mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,207 @@
+        self.methods.update(
+            {
+                "project": self.project,
+            }
+        )
+
+    def detect_tools(self, state: ModuleState, is_native: bool) -> None:
+        self.go_exe = state.find_program("go")
+
+        if not is_native and state.environment.is_cross_build():
+            state.find_program("x86_64-w64-mingw32-gcc")
+            state.find_program("x86_64-w64-mingw32-g++")
+
+    @typed_pos_args(
+        "gobuild.project",
+        str,
+        varargs=(
+            str,
+            mesonlib.File,
+            build.CustomTarget,
+            build.CustomTargetIndex,
+            build.GeneratedList,
+        ),
+    )
+    @typed_kwargs(
+        "gobuild.project",
+        CT_INPUT_KW.evolve(name="sources"),
+        KwargInfo(
+            "include_directories",
+            ContainerTypeInfo(list, (str, build.IncludeDirs)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
+        KwargInfo(
+            "gomod_tidy",
+            bool,
+            default=True,
+        ),
+    )
+    def project(
+        self,
+        state: ModuleState,
+        args: T.Tuple[
+            str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        ],
+        kwargs: ProjectKwargs,
+    ) -> ModuleReturnValue:
+        if self.go_exe is None:
+            self.detect_tools(state, kwargs["native"])
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs["sources"]))
+        )
+        self.counter += 1
+
+        go_sum_target = build.CustomTarget(
+            f"go mod tidy {self.counter}",
+            state.subdir,
+            state.subproject,
+            state.environment,
+            [
+                self.go_exe,
+                "-C",
+                "@CURRENT_SOURCE_DIR@",
+                "mod",
+                "tidy" if kwargs["gomod_tidy"] else "verify",
+            ],
+            self.interpreter.source_strings_to_files(["go.mod"]),
+            [f"go-mod-tidy-{self.counter}.log"],
+            capture=True,
+            build_always_stale=True,
+        )
+
+        build_dir_abs = Path(state.environment.get_build_dir())
+        target_dir_abs = build_dir_abs / state.backend.get_target_dir(go_sum_target)
+
+        link_dirs = {
+            "-L" + (build_dir_abs / state.backend.get_target_dir(lib)).as_posix()
+            for lib in kwargs["link_with"]
+        }
```

Comment:
The linker syntax `-L` is assuming `mingw` and/or `gcc` compilers. I am not too concerned about CGO and msvc compiler integration; most developers that I know always (cross-)compile from Linux/MacOS to windows, and never the other way around.

Reply:
Here's a place for that native keyword
```python
cc = state.environment.coredata.compilers[kwargs['native']]['c']
```
This gives you the Compiler object backing the DSL `compiler`, which gives you access to things like `has_argument` and what the argument syntax for linking is. This also has the methods to get the arguments for the C compiler you were looking for later.

While the Linux -> Windows cross makes a ton of sense, I'm hesitant to assume no one is ever going to do Windows ->Windows compilation.